File: 02 - Linked List/7.5 - Practice Day/02 - Problem 02/practiceProblem2.cpp
#include<bits/stdc++.h>
using namespace std;

class Node {
    public:
    int val;
    Node* next;
    Node(int val) {
        this->val = val;
        this->next =NULL;
    }
};

void insertAtTail(Node* &head, Node* &tail, int val){
    Node* newNode = new Node(val);
    if(head == NULL){
        head = newNode;
        tail = newNode;
        return;
    }
    tail->next = newNode;
    tail = newNode;
}

void printLinkedListInReverse(Node* tmp){
    if(tmp == NULL){
        return;
    }
    printLinkedListInReverse(tmp->next);
    cout << tmp->val << " ";
}


int main()
{
    Node* head = NULL;
    Node* tail = NULL;
    while(1){
        int val;
        cin >> val;
    if(val == -1){
        break;
    }else{
        insertAtTail(head, tail, val);
    }
    }
    Node* tmp = head;
    printLinkedListInReverse(tmp);

    return 0;
}

File: 02 - Linked List/7.5 - Practice Day/02 - Problem 02/go/practiceProblem2.go
package main

import "fmt"

type Node struct{
	val int
	next *Node
}

func insertAtTail(head **Node, tail **Node, value int){
	newNode := &Node{val: value}
	
	if *head == nil{
		*head = newNode;
		*tail = newNode
		return;
	}
	(*tail).next = newNode;
	*tail = newNode;
}

func printLinkedListInReverse(tmp **Node){
	if *tmp == nil{
		return
	}
	printLinkedListInReverse(&(*tmp).next)
	fmt.Printf("%d ", (*tmp).val);

}



func main(){

	var head *Node = nil;
	var tail *Node = nil

	var value int;
	for {
		fmt.Scan(&value)
		if value == -1{
			break;
		}
		insertAtTail(&head, &tail, value)
	}
	tmp := head;
	printLinkedListInReverse(&tmp)
}